zetaclient/chains/base/observer_internal_tracker.go

```diff
@@ -0,0 +1,102 @@
+package base
+
+import (
+	"context"
+
+	zetaerrors "github.com/zeta-chain/node/pkg/errors"
+	crosschaintypes "github.com/zeta-chain/node/x/crosschain/types"
+	"github.com/zeta-chain/node/zetaclient/config"
+	"github.com/zeta-chain/node/zetaclient/logs"
+)
+
+// GetInboundInternalTrackers returns (trackersToScan, totalCount, error)
+func (ob *Observer) GetInboundInternalTrackers(ctx context.Context) ([]crosschaintypes.InboundTracker, int) {
+	ob.mu.Lock()
+	defer ob.mu.Unlock()
+
+	var (
+		trackerCount     = 0
+		finalizedBallots = make([]string, 0)
+		trackersToScan   = make([]crosschaintypes.InboundTracker, 0)
+	)
+
+	// collect up to MaxInternalTrackersPerScan trackers
+	for ballot, tracker := range ob.internalInboundTrackers {
+		// skip those that are already finalized
+		if _, err := ob.ZetacoreClient().GetCctxByHash(ctx, ballot); err == nil {
+			finalizedBallots = append(finalizedBallots, ballot)
+			continue
+		}
+
+		trackerCount++
+		trackersToScan = append(trackersToScan, tracker)
+		if trackerCount >= config.MaxInternalTrackersPerScan {
```

Comment:
This should be done in the scanning process not in the getter function 

The getter function should return everything, the scan function take the subarray